### DISCUSSION TITLE
MRTK3 BoundsControl + ObjectManipulator for parent gameobject

### DISCUSSION THREAD

**Discussion by Spyispie120** [microsoft/MixedRealityToolkit-Unity#10998] [Q&A] [answered]
Hi all,

I'm currently running into a problem with BoundsControl + ObjectManipulator conundrum. I’m trying to create a game object hierarchy like this:

ParentGameObject (BoundControl + ObjectManipulator):
- InteractableObject:
    - …
- InteractableObject:
    - …
- MRTK Button
    - …

In MRTK2, if I have BoundsControl + ObjectManipulator in ParentGameObject, I would be able to control the bounds, manipulate the whole transform hierarchy, interact with inner child InteractableObject, override bounds for ParentGameObject and using inner MRTK Button to disable bounds if needed to…  However, with MRTK3 there seem to be a weird bug where those interaction still work in the editor, but broken after built to devices. Here are some examples:

Expectation (In editor): 
https://photos.app.goo.gl/16Rd29bAboiM4duH9

(mixed)Reality :( 
Hololens 2:
https://photos.app.goo.gl/Tf6VGYdmfNn8hTKQA

Snapdragon Space (ThinkReality A3 glasses)
https://photos.app.goo.gl/9MQVpSssdFB26U3D8

Some of the problems you can see like: some MRTK buttons work and some doesn’t (i.e. car model); Click on inner InteractableObject would interact with ParentGameObject instead of the child object, etc… To fix this, it look like disable ParentGameObject’s BoundsControl + ObjectManipulator would do the trick. The conundrum is I like to preserve BoundControl + ObjectManipulator for parentGameObject in order to control the overall bounds :(  

I just wondering if this is something that you have encountered, or is there an easy fix for this? 

Version:
MRTK3: 3.0.0-pre.10
Unity: 2021.3.4f1

Platform:
Hololens 2
Snapdragon Space

**Comment by Zee2** [comment, marked as answer]
Hi @Spyispie120 !

First thing to check: in this situation where you have several nested interactables, you'll probably want to explicitly connect the interactables (objmanips) with their colliders.

<img width="345" alt="image" src="https://user-images.githubusercontent.com/5544935/189776503-5fddc670-2442-4848-ad99-7b072c2d3b93.png">

This is a special option that XRI offers. If this collider option isn't explicitly set, the ObjectManipulator will just find the first available collider (which might already be "owned" by another interactable, like a button, or something!)

By explicitly setting the collider, you prevent these overlaps/fights/conflicts and explicitly assign the interactable <-> collider relationship.

Please let me know if this fixes it! I believe this should solve your problem because I've actually had basically the same headache myself :) I'm thinking of making this option a bit more visible because of how important it can be in certain circumstances.

Check out the `BoundsControlExamples`; I believe there's an example of a BoundsControl with buttons "underneath" it. That should serve as a good foundation for your setup (and a good sanity check!)

**Reply by Spyispie120** [reply]
Oh wow, definitely a simple fix. Tested this morning with both Hololens 2 & Snapdragon Space platform and it seem to fixed everything.

I think it's a good idea to highlight this in the documentation :) I was going through [ObjectManipulator](https://docs.microsoft.com/en-us/windows/mixed-reality/mrtk-unity/mrtk3-spatialmanipulation/packages/spatialmanipulation/object-manipulator) & [BoundsControl](https://docs.microsoft.com/en-us/windows/mixed-reality/mrtk-unity/mrtk3-spatialmanipulation/packages/spatialmanipulation/bounds-control) docs but didn't see any mention about this use case. Thanks @Zee2 !

**Reply by Zee2** [reply]
Great! We'll make sure the documentation is improved.
